File: eval/actuators/pca9685_actuator.cpp
#include "pca9685_actuator.h"

#include <fcntl.h>
#include <unistd.h>
#include <cmath>
#include <cstring>
#include <algorithm>

#if defined(__linux__)
#include <sys/ioctl.h>
#include <linux/i2c-dev.h>
#endif

namespace rvpoint {

#ifndef I2C_SLAVE
#define I2C_SLAVE 0x0703
#endif

uint64_t PCA9685Actuator::get_current_time_ns() {
    auto now = std::chrono::steady_clock::now().time_since_epoch();
    return std::chrono::duration_cast<std::chrono::nanoseconds>(now).count();
}

static float clamp_f(float val, float min_v, float max_v) {
    return std::max(min_v, std::min(val, max_v));
}

bool PCA9685Config::load_from_json(const std::string& filepath) {
    std::ifstream file(filepath);
    if (!file.is_open()) return false;
    std::string content((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());

    auto extract_float = [&](const std::string& key, float def) -> float {
        auto pos = content.find("\"" + key + "\"");
        if (pos == std::string::npos) return def;
        auto colon = content.find(':', pos);
        if (colon == std::string::npos) return def;
        return std::stof(content.substr(colon + 1));
    };

    auto extract_int = [&](const std::string& key, int def) -> int {
        auto pos = content.find("\"" + key + "\"");
        if (pos == std::string::npos) return def;
        auto colon = content.find(':', pos);
        if (colon == std::string::npos) return def;
        return std::stoi(content.substr(colon + 1));
    };

    bus = extract_int("bus", 4);
    deadband = extract_float("deadband", 0.12f);
    freq_hz = extract_float("freq_hz", 200.0f);
    return true;
}

bool PCA9685Config::save_to_json(const std::string& filepath) const {
    std::ofstream file(filepath);
    if (!file.is_open()) return false;
    file << "{\n"
         << "  \"bus\": " << bus << ",\n"
         << "  \"address\": \"" << address << "\",\n"
         << "  \"freq_hz\": " << freq_hz << ",\n"
         << "  \"mapping\": \"" << mapping << "\",\n"
         << "  \"deadband\": " << deadband << ",\n"
         << "  \"enable_watchdog\": " << (enable_watchdog ? "true" : "false") << ",\n"
         << "  \"watchdog_timeout_ms\": " << watchdog_timeout_ms << ",\n"
         << "  \"wheels\": {\n"
         << "    \"FL\": { \"motor\": " << fl.motor << ", \"invert\": " << (fl.invert ? "true" : "false") << ", \"trim\": " << fl.trim << " },\n"
         << "    \"FR\": { \"motor\": " << fr.motor << ", \"invert\": " << (fr.invert ? "true" : "false") << ", \"trim\": " << fr.trim << " },\n"
         << "    \"RL\": { \"motor\": " << rl.motor << ", \"invert\": " << (rl.invert ? "true" : "false") << ", \"trim\": " << rl.trim << " },\n"
         << "    \"RR\": { \"motor\": " << rr.motor << ", \"invert\": " << (rr.invert ? "true" : "false") << ", \"trim\": " << rr.trim << " }\n"
         << "  }\n"
         << "}\n";
    return true;
}

PCA9685Actuator::PCA9685Actuator(const PCA9685Config& config)
    : config_(config) {
    initialize_hardware();
}

PCA9685Actuator::~PCA9685Actuator() {
    shutdown_hardware();
}

bool PCA9685Actuator::initialize_hardware() {
#if defined(__linux__)
    std::string dev_path = "/dev/i2c-" + std::to_string(config_.bus);
    i2c_fd_ = open(dev_path.c_str(), O_RDWR);
    if (i2c_fd_ < 0) {
        // Try fallback bus 1
        i2c_fd_ = open("/dev/i2c-1", O_RDWR);
    }

    if (i2c_fd_ >= 0) {
        int addr = 0x60;
        try {
            addr = std::stoi(config_.address, nullptr, 0);
        } catch (...) {}
        if (ioctl(i2c_fd_, I2C_SLAVE, addr) >= 0) {
            is_simulated_ = false;
            // Mode 1: Auto-increment + allcall
            uint8_t init_mode[2] = {0x00, 0x21};
            write(i2c_fd_, init_mode, 2);
            // Mode 2: Totem pole
            uint8_t out_mode[2] = {0x01, 0x04};
            write(i2c_fd_, out_mode, 2);
        } else {
            close(i2c_fd_);
            i2c_fd_ = -1;
            is_simulated_ = true;
        }
    } else {
        is_simulated_ = true;
    }
#else
    is_simulated_ = true;
#endif

    running_ = true;
    worker_thread_ = std::thread(&PCA9685Actuator::worker_loop, this);
    return true;
}

void PCA9685Actuator::shutdown_hardware() {
    running_ = false;
    if (worker_thread_.joinable()) {
        worker_thread_.join();
    }
    emergency_brake();
    if (i2c_fd_ >= 0) {
        close(i2c_fd_);
        i2c_fd_ = -1;
    }
}

void PCA9685Actuator::hw_set_duty(int channel, float duty) {
    if (is_simulated_ || i2c_fd_ < 0 || channel < 0 || channel > 15) return;
    duty = clamp_f(duty, 0.0f, 1.0f);
    uint8_t reg_base = 0x06 + 4 * channel;

    if (duty <= 0.0f) {
        uint8_t payload[5] = {reg_base, 0, 0, 0, 0x10};
        (void)write(i2c_fd_, payload, 5);
    } else if (duty >= 1.0f) {
        uint8_t payload[5] = {reg_base, 0, 0x10, 0, 0};
        (void)write(i2c_fd_, payload, 5);
    } else {
        uint16_t off_count = static_cast<uint16_t>(duty * 4095.0f);
        uint8_t payload[5] = {
            reg_base,
            0, 0,
            static_cast<uint8_t>(off_count & 0xFF),
            static_cast<uint8_t>((off_count >> 8) & 0x0F)
        };
        (void)write(i2c_fd_, payload, 5);
    }
}

void PCA9685Actuator::hw_drive_motor(int motor_num, float duty, const std::string& dir) {
    int pwm_ch = 8, in1_ch = 10, in2_ch = 9;
    switch (motor_num) {
        case 1: pwm_ch = 8; in2_ch = 9; in1_ch = 10; break;
        case 2: pwm_ch = 13; in2_ch = 12; in1_ch = 11; break;
        case 3: pwm_ch = 2; in2_ch = 3; in1_ch = 4; break;
        case 4: pwm_ch = 7; in2_ch = 6; in1_ch = 5; break;
        default: return;
    }

    if (duty <= 0.01f) {
        hw_set_duty(pwm_ch, 0.0f);
        hw_set_duty(in1_ch, 0.0f);
        hw_set_duty(in2_ch, 0.0f);
        return;
    }

    if (dir == "forward") {
        hw_set_duty(in1_ch, 1.0f);
        hw_set_duty(in2_ch, 0.0f);
        hw_set_duty(pwm_ch, duty);
    } else {
        hw_set_duty(in1_ch, 0.0f);
        hw_set_duty(in2_ch, 1.0f);
        hw_set_duty(pwm_ch, duty);
    }
}

void PCA9685Actuator::emergency_brake() {
    emergency_stop_ = true;
    for (int m = 1; m <= 4; ++m) {
        hw_drive_motor(m, 0.0f, "forward");
    }
}

bool PCA9685Actuator::is_emergency_stopped() const noexcept {
    return emergency_stop_.load();
}

void PCA9685Actuator::reset_emergency_stop() {
    emergency_stop_ = false;
    watchdog_tripped_ = false;
    last_command_time_ns_ = get_current_time_ns();
}

void PCA9685Actuator::set_wheel_duties(float fl, float fr, float rl, float rr) {
    if (emergency_stop_) return;
    last_command_time_ns_ = get_current_time_ns();
    watchdog_tripped_ = false;

    auto apply_wheel = [&](WheelId wid, float raw) {
        const auto& cfg = config_.get_wheel_config(wid);
        bool pos = (raw >= 0.0f);
        if (cfg.invert) pos = !pos;
        float mag = std::abs(raw);
        float duty = (mag < 0.01f) ? 0.0f : clamp_f(cfg.deadband_forward + (1.0f - cfg.deadband_forward) * mag * cfg.trim, 0.0f, 1.0f);
        hw_drive_motor(cfg.motor, duty, pos ? "forward" : "reverse");
    };

    apply_wheel(WheelId::FL, fl);
    apply_wheel(WheelId::FR, fr);
    apply_wheel(WheelId::RL, rl);
    apply_wheel(WheelId::RR, rr);
}

void PCA9685Actuator::set_duty_cycles(float duty_left, float duty_right) {
    set_wheel_duties(duty_left, duty_right, duty_left, duty_right);
}

void PCA9685Actuator::worker_loop() {
    while (running_) {
        std::this_thread::sleep_for(std::chrono::milliseconds(50));
        if (!watchdog_enabled_ || emergency_stop_) continue;

        uint64_t elapsed_ms = (get_current_time_ns() - last_command_time_ns_) / 1000000;
        if (elapsed_ms > config_.watchdog_timeout_ms && !watchdog_tripped_) {
            watchdog_tripped_ = true;
            for (int m = 1; m <= 4; ++m) {
                hw_drive_motor(m, 0.0f, "forward");
            }
        }
    }
}

} // namespace rvpoint

File: demonstration/actuators/pca9685_actuator.py
#!/usr/bin/env python3
"""
pca9685_actuator.py — PCA9685 I2C Hardware PWM Motor Driver Actuator in Python.

Provides full hardware abstraction for 4-wheel Omni-Tank chassis driven by
a PCA9685 16-channel 12-bit PWM co-processor (default I2C Bus 4, Address 0x60 on Orange Pi RV2).

Features:
  - Zero external dependencies: Uses standard library Linux /dev/i2c-X with fcntl/ioctl (with smbus fallback).
  - High-precision 12-bit hardware PWM: Completely eliminates Linux CPU jitter and timer latency.
  - Direction-aware calibration: Independent forward and reverse trims and stiction deadbands per wheel.
  - Active dynamic braking and 200 ms safety watchdog.
  - Graceful mock simulation fallback when running off-target (CI/macOS/Windows).
"""

import glob
import json
import os
import signal
import sys
import threading
import time
from pathlib import Path
from typing import Any, Dict, List, Optional, Tuple


def clamp(val: float, min_val: float, max_val: float) -> float:
    return max(min_val, min(val, max_val))


# PCA9685 Registers
PCA9685_MODE1 = 0x00
PCA9685_MODE2 = 0x01
PCA9685_LED0_ON_L = 0x06
PCA9685_ALL_LED_ON_L = 0xFA
PCA9685_PRESCALE = 0xFE

MODE1_RESTART = 0x80
MODE1_AI = 0x20  # Auto-increment
MODE1_SLEEP = 0x10
MODE1_ALLCALL = 0x01
MODE2_OUTDRV = 0x04  # Totem-pole push-pull

I2C_SLAVE = 0x0703

# Standard Motor HAT PCA9685 Channel Mappings: (PWM, IN1, IN2)
ADAFRUIT_MAPPINGS = {
    1: {"pwm": 8, "in2": 9, "in1": 10},
    2: {"pwm": 13, "in2": 12, "in1": 11},
    3: {"pwm": 2, "in2": 3, "in1": 4},
    4: {"pwm": 7, "in2": 6, "in1": 5},
}

WAVESHARE_MAPPINGS = {
    1: {"pwm": 0, "in1": 1, "in2": 2},
    2: {"pwm": 5, "in1": 3, "in2": 4},
    3: {"pwm": 6, "in1": 7, "in2": 8},
    4: {"pwm": 11, "in1": 9, "in2": 10},
}


class PCA9685Actuator:
    """PCA9685 I2C 16-Channel Motor Actuator for 4-Wheel Omni-Tank Kinematics."""

    DEFAULT_CONFIG_PATH = Path("eval/actuators/pca9685_pins.json")

    def __init__(self, config_path: Optional[str | Path] = None, auto_init: bool = True):
        self.config_path = Path(config_path) if config_path else self.DEFAULT_CONFIG_PATH
        self.config: Dict[str, Any] = self._load_default_config()

        if self.config_path.exists():
            self.load_config(self.config_path)

        self._emergency_stop = False
        self._watchdog_enabled = self.config.get("enable_watchdog", True)
        self._watchdog_timeout_ms = self.config.get("watchdog_timeout_ms", 200)
        self._watchdog_tripped = False
        self._last_command_time = time.perf_counter()

        # Commanded wheel duties [-1.0, 1.0] for FL, FR, RL, RR
        self._target_duties = [0.0, 0.0, 0.0, 0.0]
        self._lock = threading.Lock()

        # Hardware handle
        self.is_simulated = False
        self.fd: Optional[int] = None
        self._smbus = None
        self._running = False
        self._worker_thread: Optional[threading.Thread] = None

        try:
            signal.signal(signal.SIGINT, self._signal_handler)
            signal.signal(signal.SIGTERM, self._signal_handler)
        except Exception:
            pass

        if auto_init:
            self.initialize_hardware()

    def _signal_handler(self, signum, frame):
        self.emergency_brake()
        self.shutdown_hardware()
        sys.exit(0)

    def _load_default_config(self) -> Dict[str, Any]:
        return {
            "bus": 4,
            "address": "0x60",
            "freq_hz": 200.0,
            "mapping": "adafruit",
            "deadband": 0.12,
            "enable_watchdog": True,
            "watchdog_timeout_ms": 200,
            "wheels": {
                "FL": {"motor": 1, "invert": False, "trim": 1.0, "deadband_forward": 0.12, "deadband_reverse": 0.12},
                "FR": {"motor": 2, "invert": False, "trim": 1.0, "deadband_forward": 0.12, "deadband_reverse": 0.12},
                "RL": {"motor": 3, "invert": False, "trim": 1.0, "deadband_forward": 0.12, "deadband_reverse": 0.12},
                "RR": {"motor": 4, "invert": False, "trim": 1.0, "deadband_forward": 0.12, "deadband_reverse": 0.12},
            },
        }

    def load_config(self, filepath: str | Path) -> bool:
        path = Path(filepath)
        if not path.exists():
            return False
        try:
            with open(path, "r", encoding="utf-8") as f:
                data = json.load(f)
            self.config.update(data)
            self.config_path = path
            return True
        except Exception as e:
            print(f"[PCA9685Actuator] Error loading {path}: {e}", file=sys.stderr)
            return False

    def save_config(self, filepath: Optional[str | Path] = None) -> bool:
        path = Path(filepath) if filepath else self.config_path
        try:
            path.parent.mkdir(parents=True, exist_ok=True)
            with open(path, "w", encoding="utf-8") as f:
                json.dump(self.config, f, indent=2)
            return True
        except Exception as e:
            print(f"[PCA9685Actuator] Error saving {path}: {e}", file=sys.stderr)
            return False

    def _auto_detect_bus_and_address(self) -> Tuple[int, int]:
        cfg_bus = int(self.config.get("bus", 4))
        raw_addr = self.config.get("address", "0x60")
        cfg_addr = int(raw_addr, 0) if isinstance(raw_addr, str) else int(raw_addr)

        # Check if configured bus node exists
        if os.path.exists(f"/dev/i2c-{cfg_bus}"):
            return cfg_bus, cfg_addr

        # Search other nodes
        nodes = sorted(glob.glob("/dev/i2c-*"))
        for node in [f"/dev/i2c-{b}" for b in [4, 1, 0, 2, 3]]:
            if os.path.exists(node):
                b = int(node.split("-")[-1])
                return b, cfg_addr

        return cfg_bus, cfg_addr

    def initialize_hardware(self) -> bool:
        bus_num, address = self._auto_detect_bus_and_address()
        dev_path = f"/dev/i2c-{bus_num}"

        if os.path.exists(dev_path):
            try:
                import fcntl

                self.fd = os.open(dev_path, os.O_RDWR)
                fcntl.ioctl(self.fd, I2C_SLAVE, address)
            except Exception:
                self.fd = None

        if self.fd is None:
            try:
                import smbus2 as smbus  # type: ignore

                self._smbus = smbus.SMBus(bus_num)
            except ImportError:
                try:
                    import smbus  # type: ignore

                    self._smbus = smbus.SMBus(bus_num)
                except ImportError:
                    self._smbus = None

        if self.fd is None and self._smbus is None:
            self.is_simulated = True
            print(f"[PCA9685Actuator] Hardware I2C /dev/i2c-{bus_num} inaccessible. Running in SIMULATION mode.")
        else:
            print(f"[PCA9685Actuator] Initialized hardware on /dev/i2c-{bus_num} (Address: 0x{address:02X})")
            self._hw_reset()
            self._hw_set_freq(float(self.config.get("freq_hz", 200.0)))

        self._running = True
        self._worker_thread = threading.Thread(target=self._watchdog_loop, daemon=True)
        self._worker_thread.start()
        return True

    def _hw_write_byte(self, reg: int, val: int) -> None:
        if self.is_simulated:
            return
        if self.fd is not None:
            os.write(self.fd, bytes([reg & 0xFF, val & 0xFF]))
        elif self._smbus is not None:
            self._smbus.write_byte_data(int(self.config.get("address", "0x60"), 0), reg & 0xFF, val & 0xFF)

    def _hw_read_byte(self, reg: int) -> int:
        if self.is_simulated:
            return 0
        if self.fd is not None:
            os.write(self.fd, bytes([reg & 0xFF]))
            d = os.read(self.fd, 1)
            return d[0] if d else 0
        elif self._smbus is not None:
            return self._smbus.read_byte_data(int(self.config.get("address", "0x60"), 0), reg & 0xFF)
        return 0

    def _hw_reset(self) -> None:
        self._hw_write_byte(PCA9685_MODE1, MODE1_AI | MODE1_ALLCALL)
        self._hw_write_byte(PCA9685_MODE2, MODE2_OUTDRV)
        time.sleep(0.005)

    def _hw_set_freq(self, freq_hz: float) -> None:
        if self.is_simulated:
            return
        prescale = int(round(25000000.0 / (4096.0 * freq_hz)) - 1.0)
        prescale = max(3, min(255, prescale))
        old_mode = self._hw_read_byte(PCA9685_MODE1)
        self._hw_write_byte(PCA9685_MODE1, (old_mode & 0x7F) | MODE1_SLEEP)
        self._hw_write_byte(PCA9685_PRESCALE, prescale)
        self._hw_write_byte(PCA9685_MODE1, old_mode)
        time.sleep(0.005)
        self._hw_write_byte(PCA9685_MODE1, old_mode | MODE1_AI | MODE1_RESTART)

    def _hw_set_duty(self, channel: int, duty: float) -> None:
        if channel < 0 or channel > 15 or self.is_simulated:
            return
        duty = max(0.0, min(1.0, float(duty)))
        reg_base = PCA9685_LED0_ON_L + 4 * channel
        if duty <= 0.0:
            payload = bytes([reg_base, 0x00, 0x00, 0x00, 0x10])
        elif duty >= 1.0:
            payload = bytes([reg_base, 0x00, 0x10, 0x00, 0x00])
        else:
            off_count = int(duty * 4095.0)
            payload = bytes([reg_base, 0x00, 0x00, off_count & 0xFF, (off_count >> 8) & 0x0F])

        if self.fd is not None:
            os.write(self.fd, payload)
        elif self._smbus is not None:
            addr = int(self.config.get("address", "0x60"), 0)
            self._smbus.write_i2c_block_data(addr, reg_base, list(payload[1:]))

    def _hw_drive_motor(self, motor_num: int, duty: float, direction: str) -> None:
        mapping_name = self.config.get("mapping", "adafruit")
        mappings = ADAFRUIT_MAPPINGS if mapping_name == "adafruit" else WAVESHARE_MAPPINGS
        if motor_num not in mappings:
            return

        pins = mappings[motor_num]
        pwm_ch = pins["pwm"]
        in1_ch = pins["in1"]
        in2_ch = pins["in2"]

        if abs(duty) < 0.01:
            self._hw_set_duty(pwm_ch, 0.0)
            self._hw_set_duty(in1_ch, 0.0)
            self._hw_set_duty(in2_ch, 0.0)
            return

        if direction == "forward":
            self._hw_set_duty(in1_ch, 1.0)
            self._hw_set_duty(in2_ch, 0.0)
            self._hw_set_duty(pwm_ch, duty)
        else:
            self._hw_set_duty(in1_ch, 0.0)
            self._hw_set_duty(in2_ch, 1.0)
            self._hw_set_duty(pwm_ch, duty)

    def shutdown_hardware(self) -> None:
        self._running = False
        if not self.is_simulated and self.fd is not None:
            try:
                # Stop all
                payload = bytes([PCA9685_ALL_LED_ON_L, 0x00, 0x00, 0x00, 0x10])
                os.write(self.fd, payload)
                os.close(self.fd)
            except Exception:
                pass
            self.fd = None

    def emergency_brake(self) -> None:
        with self._lock:
            self._emergency_stop = True
            self._target_duties = [0.0, 0.0, 0.0, 0.0]
            if not self.is_simulated:
                for m in [1, 2, 3, 4]:
                    self._hw_drive_motor(m, 0.0, "forward")

    def is_emergency_stopped(self) -> bool:
        with self._lock:
            return self._emergency_stop

    def reset_emergency_stop(self) -> None:
        with self._lock:
            self._emergency_stop = False
            self._watchdog_tripped = False
            self._last_command_time = time.perf_counter()

    def set_watchdog_enabled(self, enabled: bool) -> None:
        self._watchdog_enabled = enabled

    def set_wheel_duties(self, fl: float, fr: float, rl: float, rr: float) -> None:
        with self._lock:
            if self._emergency_stop:
                return

            self._last_command_time = time.perf_counter()
            self._watchdog_tripped = False

            raw_duties = {"FL": clamp(fl, -1.0, 1.0), "FR": clamp(fr, -1.0, 1.0),
                          "RL": clamp(rl, -1.0, 1.0), "RR": clamp(rr, -1.0, 1.0)}

            self._target_duties = [raw_duties["FL"], raw_duties["FR"], raw_duties["RL"], raw_duties["RR"]]

            wheels_cfg = self.config.get("wheels", {})
            for key, duty in raw_duties.items():
                wcfg = wheels_cfg.get(key, {"motor": 1, "invert": False, "trim": 1.0})
                motor_num = wcfg.get("motor", 1)
                invert = wcfg.get("invert", False)
                trim = wcfg.get("trim", 1.0)
                deadband = wcfg.get("deadband_forward" if duty >= 0 else "deadband_reverse", self.config.get("deadband", 0.12))

                is_positive = (duty >= 0.0)
                if invert:
                    is_positive = not is_positive

                mag = abs(duty)
                if mag < 0.01:
                    effective_duty = 0.0
                else:
                    effective_duty = deadband + (1.0 - deadband) * mag * trim
                    effective_duty = clamp(effective_duty, 0.0, 1.0)

                direction = "forward" if is_positive else "reverse"
                self._hw_drive_motor(motor_num, effective_duty, direction)

    def set_duty_cycles(self, duty_left: float, duty_right: float) -> None:
        self.set_wheel_duties(duty_left, duty_right, duty_left, duty_right)

    def _watchdog_loop(self) -> None:
        while self._running:
            time.sleep(0.05)
            if not self._watchdog_enabled or self._emergency_stop:
                continue

            with self._lock:
                elapsed_ms = (time.perf_counter() - self._last_command_time) * 1000.0
                if elapsed_ms > self._watchdog_timeout_ms and not self._watchdog_tripped:
                    # Trip watchdog and brake
                    self._watchdog_tripped = True
                    self._target_duties = [0.0, 0.0, 0.0, 0.0]
                    if not self.is_simulated:
                        for m in [1, 2, 3, 4]:
                            self._hw_drive_motor(m, 0.0, "forward")
